File: tools/question_ledger_test.py
#!/usr/bin/env python3
"""Self-running tests for tools/question_ledger.py — prove each labeling rule bites.

Run: python3 tools/question_ledger_test.py   (exit 0 = all pass).
"""
from __future__ import annotations

import json
import sys
import tempfile
from pathlib import Path

import question_ledger as ql

GOOD = {
    "id": "Q-20260708-001",
    "ts": "2026-07-08T00:00:00Z",
    "category": "CONTRARIAN",
    "target": "the witness thesis",
    "question": "Is the witness layer itself witnessed?",
    "why": "trust may have just moved up a level",
    "status": "open",
    "ticket": None,
}


def _rows(objs):
    """Turn a list of dicts (or raw strings) into read_rows-shaped tuples."""
    out = []
    for i, o in enumerate(objs, start=1):
        raw = o if isinstance(o, str) else json.dumps(o)
        try:
            obj = json.loads(raw)
        except json.JSONDecodeError:
            obj = None
        out.append((i, obj, raw))
    return out


def test_good_row_is_clean():
    assert ql.lint_rows(_rows([GOOD])) == []


def test_bad_id_rejected():
    bad = {**GOOD, "id": "Q-2026-07-08-1"}
    assert any("Q-YYYYMMDD-NNN" in p for p in ql.lint_rows(_rows([bad])))


def test_duplicate_id_rejected():
    probs = ql.lint_rows(_rows([GOOD, {**GOOD}]))
    assert any("duplicate id" in p for p in probs)


def test_unknown_category_rejected():
    bad = {**GOOD, "category": "SPICY"}
    assert any("category" in p for p in ql.lint_rows(_rows([bad])))


def test_unknown_status_rejected():
    bad = {**GOOD, "status": "parked"}
    assert any("status" in p for p in ql.lint_rows(_rows([bad])))


def test_ticketed_without_number_rejected():
    bad = {**GOOD, "status": "ticketed", "ticket": None}
    assert any("requires a positive int ticket" in p for p in ql.lint_rows(_rows([bad])))


def test_open_with_ticket_rejected():
    bad = {**GOOD, "status": "open", "ticket": 42}
    assert any("must be null" in p for p in ql.lint_rows(_rows([bad])))


def test_ticketed_with_number_ok():
    ok = {**GOOD, "status": "ticketed", "ticket": 42}
    assert ql.lint_rows(_rows([ok])) == []


def test_boolean_ticket_rejected():
    # json true is an int subclass in python — the guard must reject it.
    bad = {**GOOD, "status": "ticketed", "ticket": True}
    assert any("positive int ticket" in p for p in ql.lint_rows(_rows([bad])))


def test_extra_key_rejected():
    bad = {**GOOD, "severity": "high"}
    assert any("unexpected key" in p for p in ql.lint_rows(_rows([bad])))


def test_missing_key_rejected():
    bad = {k: v for k, v in GOOD.items() if k != "why"}
    assert any("missing key" in p for p in ql.lint_rows(_rows([bad])))


def test_non_question_rejected():
    bad = {**GOOD, "question": "This is a statement, not a question."}
    assert any("must end with '?'" in p for p in ql.lint_rows(_rows([bad])))


def test_leak_absolute_path_rejected():
    bad = {**GOOD, "question": "Why does C:\\work\\fak leak here?"}
    assert any("leaks" in p for p in ql.lint_rows(_rows([bad])))


def test_leak_email_rejected():
    bad = {**GOOD, "why": "raised by someone@example.com"}
    assert any("leaks" in p for p in ql.lint_rows(_rows([bad])))


def test_bad_json_line_rejected():
    assert any("not valid JSON" in p for p in ql.lint_rows(_rows(["{not json"])))


def test_next_id_increments():
    rows = _rows([
        {**GOOD, "id": "Q-20260708-001"},
        {**GOOD, "id": "Q-20260708-005"},
    ])
    assert ql.next_id(rows, "20260708") == "Q-20260708-006"


def test_next_id_first_of_day():
    rows = _rows([{**GOOD, "id": "Q-20260708-009"}])
    assert ql.next_id(rows, "20260709") == "Q-20260709-001"


def test_dedupe_detects_identical():
    rows = _rows([GOOD])
    hit = ql.dedupe_match(rows, "Is the witness layer itself witnessed?")
    assert hit and hit["id"] == "Q-20260708-001"


def test_dedupe_passes_novel():
    rows = _rows([GOOD])
    assert ql.dedupe_match(rows, "What breaks if the nightly target is halved?") is None


def test_stats_counts():
    rows = _rows([
        {**GOOD, "id": "Q-20260708-001", "category": "CONTRARIAN", "status": "open"},
        {**GOOD, "id": "Q-20260708-002", "category": "AFRAID", "status": "ticketed", "ticket": 7},
    ])
    s = ql.stats(rows)
    assert s["total"] == 2
    assert s["by_category"]["CONTRARIAN"] == 1
    assert s["by_status"]["ticketed"] == 1


def test_lint_cli_on_real_ledger():
    # The shipped seed ledger must itself be clean through the CLI.
    root = Path(__file__).resolve().parent.parent
    real = root / "docs" / "questions" / "asked.jsonl"
    if real.exists():
        assert ql.main(["lint", "--ledger", str(real)]) == 0


def test_absent_ledger_is_empty_not_error():
    with tempfile.TemporaryDirectory() as d:
        missing = str(Path(d) / "nope.jsonl")
        assert ql.main(["lint", "--ledger", missing]) == 0
        assert ql.main(["next-id", "--ledger", missing, "--date", "20260708"]) == 0


def _run() -> int:
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    failed = 0
    for fn in fns:
        try:
            fn()
            print(f"ok   {fn.__name__}")
        except AssertionError as e:
            failed += 1
            print(f"FAIL {fn.__name__}: {e}")
        except Exception as e:  # noqa: BLE001 — surface harness errors as failures
            failed += 1
            print(f"ERR  {fn.__name__}: {e!r}")
    print(f"\n{len(fns) - failed}/{len(fns)} passed")
    return 1 if failed else 0


if __name__ == "__main__":
    sys.exit(_run())

File: internal/questionledger/questionledger_test.go
package questionledger

import (
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// good returns a clean baseline row; tests clone and mutate it.
func good() map[string]interface{} {
	return map[string]interface{}{
		"id":       "Q-20260708-001",
		"ts":       "2026-07-08T00:00:00Z",
		"category": "CONTRARIAN",
		"target":   "the witness thesis",
		"question": "Is the witness layer itself witnessed?",
		"why":      "trust may have just moved up a level",
		"status":   "open",
		"ticket":   nil,
	}
}

// with clones good() and applies the overrides.
func with(overrides map[string]interface{}) map[string]interface{} {
	m := good()
	for k, v := range overrides {
		m[k] = v
	}
	return m
}

// rows turns dicts (or raw JSON strings) into ParseRows-shaped rows.
func rows(objs ...interface{}) []Row {
	var lines []string
	for _, o := range objs {
		if s, ok := o.(string); ok {
			lines = append(lines, s)
			continue
		}
		b, err := json.Marshal(o)
		if err != nil {
			panic(err)
		}
		lines = append(lines, string(b))
	}
	return ParseRows(strings.Join(lines, "\n"))
}

func hasProblem(probs []string, substr string) bool {
	for _, p := range probs {
		if strings.Contains(p, substr) {
			return true
		}
	}
	return false
}

func TestGoodRowIsClean(t *testing.T) {
	if probs := LintRows(rows(good())); len(probs) != 0 {
		t.Fatalf("expected clean, got %v", probs)
	}
}

func TestBadIDRejected(t *testing.T) {
	if !hasProblem(LintRows(rows(with(map[string]interface{}{"id": "Q-2026-07-08-1"}))), "Q-YYYYMMDD-NNN") {
		t.Fatal("bad id not rejected")
	}
}

func TestDuplicateIDRejected(t *testing.T) {
	if !hasProblem(LintRows(rows(good(), good())), "duplicate id") {
		t.Fatal("duplicate id not rejected")
	}
}

func TestUnknownCategoryRejected(t *testing.T) {
	if !hasProblem(LintRows(rows(with(map[string]interface{}{"category": "SPICY"}))), "category") {
		t.Fatal("unknown category not rejected")
	}
}

func TestUnknownStatusRejected(t *testing.T) {
	if !hasProblem(LintRows(rows(with(map[string]interface{}{"status": "parked"}))), "status") {
		t.Fatal("unknown status not rejected")
	}
}

func TestTicketedWithoutNumberRejected(t *testing.T) {
	bad := with(map[string]interface{}{"status": "ticketed", "ticket": nil})
	if !hasProblem(LintRows(rows(bad)), "requires a positive int ticket") {
		t.Fatal("ticketed w/o number not rejected")
	}
}

func TestOpenWithTicketRejected(t *testing.T) {
	bad := with(map[string]interface{}{"status": "open", "ticket": 42})
	if !hasProblem(LintRows(rows(bad)), "must be null") {
		t.Fatal("open with ticket not rejected")
	}
}

func TestTicketedWithNumberOK(t *testing.T) {
	ok := with(map[string]interface{}{"status": "ticketed", "ticket": 42})
	if probs := LintRows(rows(ok)); len(probs) != 0 {
		t.Fatalf("ticketed+number should be clean, got %v", probs)
	}
}

func TestBooleanTicketRejected(t *testing.T) {
	// A JSON true decodes to bool, not float64 — the guard must reject it.
	bad := with(map[string]interface{}{"status": "ticketed", "ticket": true})
	if !hasProblem(LintRows(rows(bad)), "positive int ticket") {
		t.Fatal("boolean ticket not rejected")
	}
}

func TestExtraKeyRejected(t *testing.T) {
	if !hasProblem(LintRows(rows(with(map[string]interface{}{"severity": "high"}))), "unexpected key") {
		t.Fatal("extra key not rejected")
	}
}

func TestMissingKeyRejected(t *testing.T) {
	bad := good()
	delete(bad, "why")
	if !hasProblem(LintRows(rows(bad)), "missing key") {
		t.Fatal("missing key not rejected")
	}
}

func TestNonQuestionRejected(t *testing.T) {
	bad := with(map[string]interface{}{"question": "This is a statement, not a question."})
	if !hasProblem(LintRows(rows(bad)), "must end with '?'") {
		t.Fatal("non-question not rejected")
	}
}

func TestLeakAbsolutePathRejected(t *testing.T) {
	bad := with(map[string]interface{}{"question": `Why does C:\work\fak leak here?`})
	if !hasProblem(LintRows(rows(bad)), "leaks") {
		t.Fatal("absolute path leak not rejected")
	}
}

func TestLeakEmailRejected(t *testing.T) {
	bad := with(map[string]interface{}{"why": "raised by someone@example.com"})
	if !hasProblem(LintRows(rows(bad)), "leaks") {
		t.Fatal("email leak not rejected")
	}
}

func TestBadJSONLineRejected(t *testing.T) {
	if !hasProblem(LintRows(rows("{not json")), "not valid JSON") {
		t.Fatal("bad JSON line not rejected")
	}
}

func TestNextIDIncrements(t *testing.T) {
	r := rows(
		with(map[string]interface{}{"id": "Q-20260708-001"}),
		with(map[string]interface{}{"id": "Q-20260708-005"}),
	)
	got, err := NextID(r, "20260708")
	if err != nil || got != "Q-20260708-006" {
		t.Fatalf("next-id = %q, %v; want Q-20260708-006", got, err)
	}
}

func TestNextIDFirstOfDay(t *testing.T) {
	r := rows(with(map[string]interface{}{"id": "Q-20260708-009"}))
	got, err := NextID(r, "20260709")
	if err != nil || got != "Q-20260709-001" {
		t.Fatalf("next-id = %q, %v; want Q-20260709-001", got, err)
	}
}

func TestDedupeDetectsIdentical(t *testing.T) {
	hit := DedupeMatch(rows(good()), "Is the witness layer itself witnessed?", "")
	if hit == nil || hit["id"] != "Q-20260708-001" {
		t.Fatalf("dedupe miss: %v", hit)
	}
}

func TestDedupePassesNovel(t *testing.T) {
	if hit := DedupeMatch(rows(good()), "What breaks if the nightly target is halved?", ""); hit != nil {
		t.Fatalf("novel question wrongly matched: %v", hit)
	}
}

func TestStatsCounts(t *testing.T) {
	r := rows(
		with(map[string]interface{}{"id": "Q-20260708-001", "category": "CONTRARIAN", "status": "open"}),
		with(map[string]interface{}{"id": "Q-20260708-002", "category": "AFRAID", "status": "ticketed", "ticket": 7}),
	)
	s := Stats(r)
	if s.Total != 2 || s.ByCategory["CONTRARIAN"] != 1 || s.ByStatus["ticketed"] != 1 {
		t.Fatalf("stats wrong: %+v", s)
	}
}

func TestLintCLIOnRealLedger(t *testing.T) {
	real := filepath.Join("..", "..", "docs", "questions", "asked.jsonl")
	if _, err := os.Stat(real); err != nil {
		t.Skipf("no seed ledger at %s", real)
	}
	var out, errb bytes.Buffer
	if rc := Run(&out, &errb, []string{"lint", "--ledger", real}, "20260708", nil); rc != 0 {
		t.Fatalf("real ledger not clean: rc=%d out=%s err=%s", rc, out.String(), errb.String())
	}
}

func TestAbsentLedgerIsEmptyNotError(t *testing.T) {
	missing := filepath.Join(t.TempDir(), "nope.jsonl")
	var out, errb bytes.Buffer
	if rc := Run(&out, &errb, []string{"lint", "--ledger", missing}, "20260708", nil); rc != 0 {
		t.Fatalf("absent ledger lint rc=%d", rc)
	}
	out.Reset()
	errb.Reset()
	if rc := Run(&out, &errb, []string{"next-id", "--ledger", missing, "--date", "20260708"}, "20260708", nil); rc != 0 {
		t.Fatalf("absent ledger next-id rc=%d", rc)
	}
	if got := strings.TrimSpace(out.String()); got != "Q-20260708-001" {
		t.Fatalf("next-id on empty = %q; want Q-20260708-001", got)
	}
}

func TestDedupeCheckExitCode(t *testing.T) {
	// Round-trips through Run so the exit-3 dedupe contract is covered.
	dir := t.TempDir()
	ledger := filepath.Join(dir, "asked.jsonl")
	b, _ := json.Marshal(good())
	if err := os.WriteFile(ledger, b, 0o644); err != nil {
		t.Fatal(err)
	}
	var out, errb bytes.Buffer
	rc := Run(&out, &errb, []string{"dedupe-check", "--ledger", ledger, "--question", "Is the witness layer itself witnessed?"}, "20260708", nil)
	if rc != 3 {
		t.Fatalf("dedupe hit rc=%d; want 3 (out=%s)", rc, out.String())
	}
	out.Reset()
	rc = Run(&out, &errb, []string{"dedupe-check", "--ledger", ledger, "--question", "An entirely unrelated novel question?"}, "20260708", nil)
	if rc != 0 {
		t.Fatalf("novel dedupe rc=%d; want 0", rc)
	}
}

func TestEnsureLabelPrints(t *testing.T) {
	var out, errb bytes.Buffer
	if rc := Run(&out, &errb, []string{"ensure-label"}, "20260708", nil); rc != 0 {
		t.Fatalf("ensure-label rc=%d", rc)
	}
	if !strings.Contains(out.String(), "gh label create "+Label) {
		t.Fatalf("ensure-label print = %q", out.String())
	}
}
